### PR TITLE
refactor(ivy): introduce the 'core' package and split apart NgtscProgram

### PR DESCRIPTION
**Pull request by alxhub**
Previously, NgtscProgram lived in the main @angular/compiler-cli package
alongside the legacy View Engine compiler. As a result, the main package
depended on all of the ngtsc internal packages, and a significant portion of
ngtsc logic lived in NgtscProgram.

This commit refactors NgtscProgram and moves the main logic of compilation
into a new 'core' package. The new package defines a new API which enables
implementers of TypeScript compilers (compilers built using the TS API) to
support Angular transpilation as well. It involves a new NgCompiler type
which takes a ts.Program and performs Angular analysis and transformations,
as well as an NgCompilerHost which wraps an input ts.CompilerHost and adds
any extra Angular files.

Together, these two classes are used to implement a new NgtscProgram which
adapts the legacy api.Program interface used by the View Engine compiler
onto operations on the new types. The new NgtscProgram implementation is
significantly smaller and easier to reason about.

The new NgCompilerHost replaces the previous GeneratedShimsHostWrapper which
lived in the 'shims' package.

A new 'resource' package is added to support the HostResourceLoader which
previously lived in the outer compiler package.

As a result of the refactoring, the dependencies of the outer
@angular/compiler-cli package on ngtsc internal packages are significantly
trimmed.

This refactoring was driven by the desire to build a plugin interface to the
compiler so that tsc_wrapped (another consumer of the TS compiler APIs) can
perform Angular transpilation on user request.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+
+/**
+ * A host backed by a build system which has a unified view of the module namespace.
+ *
+ * Such a build system is able to map be supports the `moduleNameToFileName` method provided by
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/core/api.ts]: I don't understand this comment, `able to map be supports the` seems incorrect.

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+      }
+    }
+
+    if (file === undefined) {
+      return this.diagnostics;
+    } else {
+      return this.diagnostics.filter(diag => {
+        if (diag.file === file) {
+          return true;
+        } else if (isTemplateDiagnostic(diag) && diag.componentFile === file) {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: This would be worth having in a unit test, I don't think we have such a test at the moment.

**Review by JoostK — APPROVED** [review]
LGTM, with some more minor comments

File: packages/compiler-cli/src/ngtsc/core/src/host.ts [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+/**
+ * A wrapper around `ts.CompilerHost` (plus any extension methods from `ExtendedTsCompilerHost`).
+ *
+ * In order for a consumer to include Angular compilation in their TypeScript compiler, the
+ * `ts.Program` must be created with a host that adds Angular-specific files (e.g. factories,
+ * summaries, the template type-checking file, etc) to the compilation. `NgCompilerHost` is the
+ * host implementation which supports this.
+ *
+ * The interface implementations here ensure that `NgCompilerHost` fully delegates to
+ * `ExtendedTsCompilerHost` methods whenever present.w
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/core/src/host.ts]:
```suggestion
 * `ExtendedTsCompilerHost` methods whenever present.
```

File: packages/compiler-cli/src/ngtsc/core/src/host.ts [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  }
+
+  getSourceFile(
+      fileName: string, languageVersion: ts.ScriptTarget,
+      onError?: ((message: string) => void)|undefined,
+      shouldCreateNewSourceFile?: boolean|undefined): ts.SourceFile|undefined {
+    for (let i = 0; i < this.shims.length; i++) {
+      const generator = this.shims[i];
+      // TypeScript internal paths are guaranteed to be POSIX-like absolute file paths.
+      const absoluteFsPath = resolve(fileName);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/core/src/host.ts]: We could use `absoluteFrom` rather than `resolve` here, given that TS internal paths are guaranteed to be absolute.

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 5]
```diff
@@ -383,56 +157,39 @@ export class NgtscProgram implements api.Program {
...
+              this.compiler.incrementalDriver.recordSuccessfulEmit(writtenSf);
             }
           }
+
+          // If Closure annotations are being produced, tsickle should be adding `@nocollapse` to
+          // any static fields present. However, tsickle doesn't yet handle synthetic fields added
+          // during other transformations, so this hack is in place to ensure Ivy definitions get
+          // properly annotated, pending an upstream fix in tsickle.
+          //
+          // TODO(alxhub): remove when tsickle properly annotates synthetic fields.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/program.ts]: Is there a tsickle issue to track for this?

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+    let importRewriter: ImportRewriter;
+    if (coreImportsFrom !== null) {
+      importRewriter = new R3SymbolsImportRewriter(coreImportsFrom.fileName);
+    } else {
+      importRewriter = new NoopImportRewriter();
+    }
+
+    // The formatter really doesn't appreciate these casts, and no amount of begging or pleading
+    // with parentheses seems to convince it to leave these lines alone. Thus, it must be disabled
+    // for each one.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: What casts?

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+          // When a pipe in scope is updated, the component needs to be recompiled as e.g. the
+          // pipe's name may have changed.
+          depGraph.addTransitiveDependency(file, pipe.ref.node.getSourceFile());
+        }
+      }
+    }
+    this.perfRecorder.stop(recordSpan);
+  }
+
+  private scanForMwp(sf: ts.SourceFile): void {
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: Probably not part of this PR but it seems strange for this to be in the compiler rather than inside a trait? Am I missing something?

**Review by petebacondarwin — APPROVED** [review]
Nice cleanup! Lots of minor NITS. Nothing to get too worried about.

File: packages/compiler-cli/src/ngtsc/core/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# What is the 'core' package?
+
+This package contains the core functionality of the Angular compiler. It provides APIs for the implementor of a TypeScript compiler to provide Angular compilation as well.
+
+Its supports the 'ngc' command-line tool and the Angular CLI (via the `NgtscProgram`).
+
+# Angular compilation
+
+Angular compilation involves the translation of Angular decorators into static definition fields. At build time, this is done during the overall process of TypeScript compilation, where TypeScript code is type-checked and then downleveled to JavaScript code. Along the way, diagnostics specific to Angular can also be produced.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/README.md]: Line breaks in Markdown are semantic, so long lines are actually what we want here. Otherwise we're breaking the formatting.

File: packages/compiler-cli/src/ngtsc/core/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+2. That `ts.CompilerHost` is wrapped in an `NgCompilerHost`, which adds Angular specific files to the compilation.
+3. A `ts.Program` is created from the `NgCompilerHost` and its augmented set of root files.
+4. An `NgCompiler` is created using the `ts.Program`.
+5. Diagnostics can be gathered from the `ts.Program` as normal, as well as from the `NgCompiler`.
+6. Prior to `emit`, `NgCompiler.prepareEmit` is called to retrieve the Angular transformers which need to be fed to `ts.Program.emit`.
+7. `emit` is called on the `ts.Program` with the Angular transformers from above, which produces JavaScript code with Angular extensions.
+
+## Asynchronous compilation
+
+In some compilation environments (such as the Webpack-driven compilation inside the Angular CLI), various inputs to the compilation are only producible in an asynchronous fashion. For example, SASS compilation of `styleUrls` that link to SASS files requires spawning a child Webpack compilation. To support this, Angular has an asynchronous interface for loading such resources.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/README.md]: Nope! The actual TS side is (particularly the module resolution hooks that ngcc cares about). But the Angular analysis is async, because producing `styleUrls` or `templateUrl` contents might require running through another Webpack invocation, which I believe is accomplished by forking the webpack-ing process?

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+ * certain build systems (such as Bazel).
+ */
+export interface UnifiedModulesHost {
+  /**
+     * Converts a file path to a module name that can be used as an `import ...`
+     * I.e. `path/to/importedFile.ts` should be imported by `path/to/containingFile.ts`.
+     */
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: I believe `path/to/containingFile.ts` is supposed to be the example, but I've clarified it to be a lot more clear.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+
+/**
+ * A host backed by a build system which has a unified view of the module namespace.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: It can be used in other contexts :) actually, the compiler has an internal interface named `FileToModuleHost` which is identical to this one. I've modified the PR to unify them (I like the name `UnifiedModulesHost` anyway as it's a lot more clear about what the intent of the interface is - a unified view of the module namespace.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+   */
+  getModifiedResourceFiles?(): Set<string>|undefined;
+}
+
+/**
+ * A `ts.CompilerHost` interface which supports some number of optional methods in addition to the
+ * core interface.
+ */
+export interface ExtendedTsCompilerHost extends ts.CompilerHost, Partial<ResourceHost>,
+    Partial<UnifiedModulesHost>, ts.CompilerHost {}
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Lol. It's _really_ extended, okay?

Fixed.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+   * For maximum type-checking, set this to `true`, and set `strictTemplates` to `true`.
+   *
+   * It is an error for this flag to be `false`, while `strictTemplates` is set to `true`.
+   */
+  fullTemplateTypeCheck?: boolean;
+
+  // Whether to generate a flat module index of the given name and the corresponding
+  // flat module metadata. This option is intended to be used when creating flat
+  // modules similar to how `@angular/core` and `@angular/common` are packaged.
+  // When this option is used the `package.json` for the library should referred to the
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: This is copy-pasted from the old transformers/api.ts. But fixed :)

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+  // generated flat module index instead of the library index file. When using this
+  // option only one .metadata.json file is produced that contains all the metadata
+  // necessary for symbols exported from the library index.
+  // In the generated .ngfactory.ts files flat module index is used to import symbols
+  // includes both the public API from the library index as well as shrowded internal
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Same, but fixed.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+  // symbols.
+  // By default the .ts file supplied in the `files` files field is assumed to be
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Same, but fixed.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+  // library index. If more than one is specified, uses `libraryIndex` to select the
+  // file to use. If more than on .ts file is supplied and no `libraryIndex` is supplied
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: ^

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+  // `package.json`'s `module` field would be `"index.js"` and the `typings` field would
+  // be `"index.d.ts"`.
+  flatModuleOutFile?: string;
+
+  // Preferred module id to use for importing flat module. References generated by `ngc`
+  // will use this module name when importing symbols from the flat module. This is only
+  // meaningful when `flatModuleOutFile` is also supplied. It is otherwise ignored.
+  flatModuleId?: string;
+
+  // Always report errors when the type of a parameter supplied whose injection type cannot
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: ^

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+   * This flag currently controls a couple aspects of template type-checking, including
+   * whether embedded views are checked.
+   *
+   * For maximum type-checking, set this to `true`, and set `strictTemplates` to `true`.
+   *
+   * It is an error for this flag to be `false`, while `strictTemplates` is set to `true`.
+   */
+  fullTemplateTypeCheck?: boolean;
+
+  // Whether to generate a flat module index of the given name and the corresponding
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Okay, okay, okay, haha. I cleaned up all the comments.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+   * Tells the compiler to generate definitions using the Render3 style code generation.
+   * This option defaults to `true`.
+   *
+   * Acceptable values are as follows:
+   *
+   * `false` - run ngc normally
+   * `true` - run the ngtsc compiler instead of the normal ngc compiler
+   * `ngtsc` - alias for `true`
+   *
+   * @publicApi
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Some mysteries are better left alone.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+  /**
+   * Whether to check the type of a binding to a directive/component input against the type of the
+   * field on the directive/component.
+   *
+   * For example, if this is `false` then the expression `[input]="expr"` will have `expr` type-
+   * checked, but not the assignment of the resulting type to the `input` property of whichever
+   * directive or component is receiving the binding. If set to `true`, both sides of the assignment
+   * are checked.
+   *
+   * Defaults to `false`, even if "fullTemplateTypeCheck" is set.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: This field is `strictInputTypes` :-P

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+   */
+  strictContextGenerics?: boolean;
+}
+
+/**
+ * Options which control behavior useful for "monorepo" build cases using Bazel (such as the
+ * internal Google monorepo, g3).
+ */
+export interface BazelAndG3Options {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: I think it's fine. G3 has a lot of special casing on top of Bazel (e.g. the Closure options).

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+  generateDeepReexports?: boolean;
+}
+
+/**
+ * Options related to i18n compilation support.
+ */
+export interface I18nOptions {
+  // Locale of the imported translations
+  i18nInLocale?: string;
+}
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Done. I copied over a few other options which do seem to be used in Ivy.

Since it's referenced, I'm going to leave it alone in this refactor.

File: packages/compiler-cli/src/ngtsc/core/api.ts [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+   * This is only active when building with `enableIvy: true`. Defaults to `true`.
+   */
+  compileNonExportedClasses?: boolean;
+
+  // Whether to remove blank text nodes from compiled templates. It is `false` by default starting
+  // from Angular 6.
+  preserveWhitespaces?: boolean;
+}
+
+export interface LazyRoute {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/api.ts]: Copied from old view engine docs. I don't even know what these fields do, man...

File: packages/compiler-cli/src/ngtsc/core/src/host.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/src/host.ts]: Done!

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts (hunk 6 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: Joost had me refactor them away. Comment deleted!

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts (hunk 7 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: `Trait` is an operation with a pretty specific shape - match a decorator or some other aspect of the class, extract information from it in local & global steps, and then add some static fields to the class (mimicking the execution of said decorator at runtime).

Other forms of static analysis & mutation, like detecting static methods or functions which return `ModuleWithProviders`, don't really fit this model. We could create a similar kind of abstraction (maybe upstream ngcc's `Migration` concept?).

File: packages/compiler-cli/src/ngtsc/core/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/core/README.md]: In github they're not: https://github.github.com/gfm/#paragraphs

You need an empty blank line to indicate an end of paragraph.

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts (hunk 7 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: OK - may be for another PR :-)

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: Agreed. This was a little tricky to do at this level of abstraction - I think we could benefit from some testing utilities here eventually. Still, the test looks pretty reasonable.
